### ISSUE TITLE
Failed BMI provisions can lead to an inconsistent state in BMI

### ISSUE THREAD

**Issue by sirushtim** [CCI-MOC/ims#57]
ubuntu@test-ceph:~/ims/ims$ bmi pro bmi_infra some_bmi_node bmi_image bmi_network 1
bmi_image not found

ubuntu@test-ceph:~/ims/ims$ bmi dpro bmi_infra some_bmi_node bmi_network 1
Got status code 400 from HaaS with message : The network is not attached to the nic.

The issue is likely to be at the tear-down/roll-back process for failed BMI provisions.

**Comment by chemistry-sourabh** [comment]
In most of the cases BMI is not left in an inconsistent state due to the rollback process. But it will happen if the rollback process is screwed up!!

**Comment by chemistry-sourabh** [comment]
It is important to identify what all are the cases where the rollback process fails and possible contingency plans. Also I always wanted to make a reset operation which sort of brings BMI to a consistent state. If we shift to postgres we can probably rollback to the previous commit ?

**Comment by sirushtim** [comment]
> In most of the cases BMI is not left in an inconsistent state due to the rollback process. But it will happen if the rollback process is screwed up!!

Right, BMI does go to an inconsistent state because the roll back process is inherently flawed.

> It is important to identify what all are the cases where the rollback process fails and possible contingency plans. Also I always wanted to make a reset operation which sort of brings BMI to a consistent state. If we shift to postgres we can probably rollback to the previous commit ?

Sure, besides changing a row in a database, we also need to check for cases where we handle stale resources that BMI creates and is not handled appropriately.

Also, for what it's worth, trying to roll back to a previous commit is going to create a long running transaction in the database by sqlalchemy while we wait for the creation of other resources making it inefficient. I'd much rather would like to see a state machine model setting the node to an error state should something happen while we create and letting the user know what happened instead of rolling back and deleting the row entirely. IMO, that's a better user experience than just deleting the node if something went wrong.

**Comment by chemistry-sourabh** [comment]
Your idea sounds more like what openstack does which makes sense and will also reduce a lot of code. But in openstack if the delete of something like an image results in failure., the image doesn't go into a failure state. 

Also if the provision fails and it ends up in the error state how will the user undo it ? By deproing it ? in that case our depro should be made more robust. What if the depro fails ? Should the user continue deproing or proing ? To avoid these confusions it is better to do a reset which removes the node from the error state.

I like this idea so this what I propose.
pro just tries to provision and if fails then ends up in an error state showing the error.
dpro is exactly like pro but tries to dpro.
Both commands will fail if node is already in error state.
reset will bring the node to the dproed state from whatever state it is in.
I feel this makes the commands atomic like a DB.

This needs to be properly discussed so let me know what you guys think ?

**Comment by sirushtim** [comment]
> Your idea sounds more like what openstack does which makes sense and will also reduce a lot of code. But in openstack if the delete of something like an image results in failure., the image doesn't go into a failure state.

AFAIK, that's not true. Sometimes, there's a DELETE_FAILED state to indicate to the user when the delete of a physical resource in the infrastructure does not work. This gives a chance for the user to check what's happening with the infrastructure before changing the state of the database.

> Also if the provision fails and it ends up in the error state how will the user undo it ? By deproing it ? in that case our depro should be made more robust. What if the depro fails ? Should the user continue deproing or proing ? To avoid these confusions it is better to do a reset which removes the node from the error state.

This is good except I'd like to keep the API minimalistic. We shouldn't add a new API call for every corner case though.

What we should just have is besides an error state is:

bmi pro
bmi dpro
bmi dpro --force (document clearly with a cautionary warning that this will be at user's own risk and there may be stale resources left behind)

**Comment by chemistry-sourabh** [comment]
@CCI-MOC/bmi-breakers  I recommend a vote on the API here

**Comment by chemistry-sourabh** [comment]
@CCI-MOC/bmi-breakers this is the next bug that will improve stability for the next release. So it is better to decide on the approach

**Comment by chemistry-sourabh** [comment]
Also the rollback thing when RegisterFailedException should be addressed with this.

**Comment by apoorvemohan** [comment]
@chemistry-sourabh Did you created a PR for this?

**Comment by chemistry-sourabh** [comment]
Never decided on what to do

**Comment by chemistry-sourabh** [comment]
@CCI-MOC/bmi-breakers how to solve this ?

**Comment by naved001** [comment]
@chemistry-sourabh I think we need to identify cases where a provision can fail (like #91) and then fix those cases 1 by 1.

**Comment by sirushtim** [comment]
We should definitely create a state machine model in the future to make things easier.

For now, the solution to solve this particular bug however is to simply not assume that a 400 from HaaS is an error which is what the code in operations.py does ie we should simply move on if the network is not attached to the NIC in the deprovision process which is what deprovision would've done perhaps in a previous deprovision call.

Except this isn't straightforward since the deprovision method in operations.py catches all exceptions. The code would have to localise try and except to each client call. The fact that we have a global try and except is making operations coupled and problematic.

So, fixing it could go about in the following fashion:

- Localise try and except for interface calls. Make a separate method for each interface call and then call them sequentially in operations.py.

- When we catch exceptions, check if the exception is appropriate. Take above example, if when deprovisioning, HaaS tells us the network is already detached from the nic, that's the expected result, so instead of raising an exception, we ignore that exception and move onto the next step in deprovision. If it is actually an error we can't deal with, re-raise it.

^ This also makes it easier to move to a state machine model FWIW.

**Comment by chemistry-sourabh** [comment]
I would suggest doing it properly so that we dont have to come back to this. So implementing the state machine model or some other solution is better instead of a temp solution.

Right now we try to rollback depending on the exception hit, but the rollback could fail which also causes issues.

For the detach thing I recommend catching the 400 in HIL driver as it will streamline stuff better. Also similarly for attach also.

We have similar situations for db and ceph also. We can catch the Integrity Error and ImageExists Error and do nothing. 

So pro will become one big try catch block and we will return error if something else happens we will not rollback.

For dpro I feel it should only dpro and not rollback as I feel it is misleading in terms of state machine model. dpro or pro should print saying that rollback using rollback command.

**Comment by sirushtim** [comment]
@chemistry-sourabh This isn't a temp model. It's rather an iterative one. Perhaps you're missing out on something. A state machine will just be plugged into the existing model once the localisation is done.

> For the detach thing I recommend catching the 400 in HIL driver as it will streamline stuff better.
> similarly for attach also.
> We have similar situations for db and ceph also. We can catch the Integrity Error and ImageExists > Error and do nothing.

We already do catch 400 in HIL. We'd just need to make sure we ignore certain exceptions thrown by the HIL client in operations.py as and when appropriate.

> So pro will become one big try catch block and we will return error if something else happens we   > will not rollback.
> For dpro I feel it should only dpro and not rollback as I feel it is misleading in terms of state             > machine model. dpro or pro should print saying that rollback using rollback command.

We will have state changes for a node in each localised try-catch block to set the state to active/rollback/error.

Basically, I'd like code to be as decoupled as much as possible first as that gives us more flexibility and it's easier to modify in the future. That means no BIG try-catch blocks encapsulating multiple operations. :) 

Then, we move onto implementing a state machine model where each phase sets the node to an error/active or rollback state.

**Comment by chemistry-sourabh** [comment]
We are not catching 400 error, but raising it as UnknownException. We can catch the UnknownException (or a variant of HILException) in the detach/attach code and not raise it.

I dont understand on how are you planning to crash it if something wrong happens without a big try catch. If attach fails due to some serious exception we need to exit immediately with the proper state set.

**Comment by sirushtim** [comment]
> We are not catching 400 error, but raising it as UnknownException. We can catch the UnknownException (or a variant of HILException) in the detach/attach code and not raise it.

Ah yes, I meant we handled 400 in HIL client. What I meant was we should definitely ignore the specific 400 in operations.py and not in the HIL driver.

> I dont understand on how are you planning to crash it if something wrong happens without a big try catch. If attach fails due to some serious exception we need to exit immediately with the proper state set.

Each operation would have it's own try catch that will set states accordingly. We'll have to check at every interface operation, perhaps use a decorator to check if we're in the correct state and if not, exit immediately with the last known error stored in the state object.

**Comment by chemistry-sourabh** [comment]
Why in operations.py and not in HIL Driver ? I feel it will reduce redundancy.

Your states thing doesnt make sense to me. These decorated methods will again be called in pro or dpro right ? We will need a try catch there to skip the sequence right ?

**Comment by sirushtim** [comment]
> Why in operations.py and not in HIL Driver ? I feel it will reduce redundancy.

because operations.py will decide which exception to ignore for what call ie dpro and pro. If not, you could create a separate method in the HIL client that deals with it. So, HIL.delete_network can handle and ignore the exception if the network is already deleted. That however is a minor detail.

> Your states thing doesnt make sense to me. These decorated methods will again be called in pro or dpro right ? We will need a try catch there to skip the sequence right ?

Yeah. We'll have one simple try-catch just to exit the sequence. Minor detail again.

It's simple, basically, interfaces are responsible for setting the deploy state of the node. We don't rely on the sequence of the deploy code to handle rollback. We execute the same sequence everytime. If it already exists or is already deleted, we move on.

**Comment by chemistry-sourabh** [comment]
So if the state is correct we wont check whether it is actually there in db, ceph or tgt ? I feel that is a bad idea.

We can just modify the interface  functions to ignore the duplicate exceptions and use the simple try catch. This will result in same effect as your decorated methods and in the simple catch we just set the state to ERROR and exit with message.

**Comment by sirushtim** [comment]
We need to trust the interface to take care of the state.

Yeah, I think you finally got the idea.

**Comment by ravisantoshgudimetla** [comment]
So, what is the plan to save the state of node here? Where are we going to persist it?

**Comment by chemistry-sourabh** [comment]
As per my understanding we don't need to persist anything additional. Just a variable called state in BMI class that stores whether in WORKING or in ERROR state.

**Comment by sirushtim** [comment]
The database will have the state of the node. Our node_get REST API call will just get the database entry and we'll know the state. Everytime we want to change states, we update the state and if necessary have a last_error_message for the user.

**Comment by chemistry-sourabh** [comment]
I forgot about the per node state thing, but we wanted to avoid adding node stuff into the db. Is there any other solution for this ?

**Comment by sirushtim** [comment]
Why do you want to avoid adding node data in the db?

**Comment by chemistry-sourabh** [comment]
1 word @ravisantoshgudimetla :P

This was in the 1st week of my internship, so didn't question him about it. If I remember right he believed that BMI should store only image related data and all the node related data should be in HIL.

**Comment by sirushtim** [comment]
BMI keeps track of the deployed state of the node, so it's fair to have the node ID of a provisioned node with its deploy state in the BMI database. It's a way of tracking what the provisioning engine is doing at any given moment to a particular node. This piece of information which wouldn't be appropriate to have in HIL.

**Comment by ravisantoshgudimetla** [comment]
> 1 word @ravisantoshgudimetla :P

@chemistry-sourabh Wait, that's not a single word and as if you were listening to me from day 1 :P

Anyways, @sirushtim yes, I am against saving node information(like vlan, project etc) but node status is fine, in fact currently, we are doing the same thing, we are returning user the status of provision/deprovision operations but we are doing it in memory. 

One thing that I expect you guys to come up with is a data model(how to sync it with external system from which we get the node info), changes to existing implementation and how it impacts the existing system. In fact, since @sirushtim has previous openstack dev experience, it would be really helpful if he can write a proposal of the format that we were discussing in #96. LMK, if you have any questions.

**Comment by chemistry-sourabh** [comment]
Ok if we are creating a node table, should we connect it to the image table and project table also ? The relationships will show that the node is under a project and provisions a particular image.

This will remove the restriction that users cannot name the snapshots as node names.

**Comment by chemistry-sourabh** [comment]
I started coding this. We probably won't need a node db as how the implementation is going and as a side effect will also need to make other operations like create_snapshot and import also state based.

If anybody is curious about the approach, I am sought doing it the way puppet does installations like if it is already attached move to next step, if clone exists move to next step, and if something screws up crash. Already done with provision (In a Basic way) works well!! Need to make sure all the checks in each step are the best possible.

**Comment by ravisantoshgudimetla** [comment]
I did not understand the  crash part, what is going to happen if the provisioning for some reason bmi server crashes in the middle? If you don't persist it, how are you planning to restore the state?

**Comment by apoorvemohan** [comment]
@chemistry-sourabh The changes you are proposing will take "a lot of time" (2-3 weeks) including the test cases for them. For 0.5 we will go with fixing uncaught exceptions.

**Comment by chemistry-sourabh** [comment]
@ravisantoshgudimetla The code will start from the beginning. It will try to attaching to hil and it it already is then it will move to the next step, if everything goes alright then the provision should succeed or crash at the same spot or in a new place.

I feel this is better as the admin may manually need to reset the state if we store it in the db.

**Comment by ravisantoshgudimetla** [comment]
>The code will start from the beginning. It will try to attaching to hil and it it already is then it will move to the next step

Its kind of easy to do for HIL, what about Ceph, are you planning to clean it up before provisioning again, if yes, you need to have a state in DB.(saying ceph operation done or HIL operation succeeded etc).

**Comment by ravisantoshgudimetla** [comment]
> The changes you are proposing will take "a lot of time" (2-3 weeks) including the test cases for them. For 0.5 we will go with fixing uncaught exceptions and rolling back any persisted changes.

I agree with @apoorvemohan. Would like to hear @sirushtim's thoughts as well.

**Comment by chemistry-sourabh** [comment]
@ravisantoshgudimetla Why do you need to clean it up ? If the clone exists we are good right ?

**Comment by ravisantoshgudimetla** [comment]
The state might have changed by the time. So, say if at state1 - HIL & Ceph provisioning succeeded and it failed at PXE template file creation & BMI server crashed, how do we ensure that once BMI server comes up, it starts from PXE template creation?

**Comment by chemistry-sourabh** [comment]
@ravisantoshgudimetla We don't. It starts from beginning and checks if everything is ok and moves on to PXE template. The admin may do something that may result in change of state. He/She will manually need to change the state and if the state set is wrong or some other stuff is not done right more errors erupt. I feel storing state will make it more difficult to make it robust.

It is the same old provision and deprovision, only difference is that it won't crash if the step is already done and there is no rollback.

I already finished implementing this for pro and depro. Just trying to write it cleaner.

**Comment by chemistry-sourabh** [comment]
@apoorvemohan if you just want the exception thing fixed then please raise a separate issue and assign it to @naved001. He showed me a snippet that should fix the RegisterFailException, but it won't completely fix the inconsistent state thing.

**Comment by apoorvemohan** [comment]
@chemistry-sourabh Will be doing that tomorrow with him. Yes I was there when he sent you the snippet.

**Comment by ravisantoshgudimetla** [comment]
> checks if everything is ok and moves on to PXE template

How does BMI server know that it is ok after it crashed? Thats the question.

**Comment by chemistry-sourabh** [comment]
@ravisantoshgudimetla let me give an example.
Lets say we do a pro and it crashes at PXE Template stuff due to some permissions error (Last step).

So now lets say the admin fixes the permissions and restarts BMI.

The user runs the pro command again. Then bmi does this,
1. Tries to attach to network in HIL, gets attached exception then moves to db entry.
2. Tries to insert new row into db, gets IntegrityError then does an additional check to check parent of image in db if matches then moves to on.
3. Tries to clone in ceph, will get image exists error then does a similar check like above to make sure parent is right then moves on.
4. Rewrites tgt config files and does --execute. (Basically remounts)
5. Creates PXE template files (which should succeed since permissions were fixed).
6. Shows Success!!

If any of this steps screws up will crash and show appropriate error.
As per I my understanding you won't need to store state for this as state is being recomputed each time. I feel this is better than storing state as it may go stale.

**Comment by naved001** [comment]
@chemistry-sourabh and I discussed that we could dpro similarly. Eg; See if the network has been detached, if yes, then move on to reverting the next thing. So I was imagining that if we fail a provision (let's say due to some typo) and want to get back to step 0 (basically reset everything about that node) we could just run deprovision on it. Correct me if I am wrong @chemistry-sourabh

**Comment by chemistry-sourabh** [comment]
Ya exactly

**Comment by ravisantoshgudimetla** [comment]
>1.)Tries to attach to network in HIL, gets attached exception then moves to db entry.
>2.)Tries to insert new row into db, gets IntegrityError then does an additional check to check parent of image in db if matches then moves on.

What if I give a new snapshot name, after this.(I mean, I technically want to provision a node with a new snapshot that I have). Should I deprovision it and then provision it again or there is a automatic rollback?

**Comment by chemistry-sourabh** [comment]
Yes what you said. There is no automatic rollback as it is part of the problem.

**Comment by ravisantoshgudimetla** [comment]
I thought blind rollbacks are problematic and you wanted to do introduce some intelligence into it. Isn't it the case?

**Comment by chemistry-sourabh** [comment]
Lol no. The issue with the rollback is when they crash. The solution is to simplify.

**Comment by chemistry-sourabh** [comment]
Also please reply asap as I want to finish this

**Comment by sirushtim** [comment]
@chemistry-sourabh When will we have a scenario with a missing db entry and the image is still present in ceph? Through what sequence of events?

**Comment by chemistry-sourabh** [comment]
@sirushtim db could get corrupted or the admin screws up.

**Comment by sirushtim** [comment]
@chemistry-sourabh In that case, I would vote for leaving the state as it is(to give a chance for the admin to rectify the state), throw a critical warning that the db is in an inconsistent state, and show FAILED because it is not a successful deprovision.

**Comment by naved001** [comment]
We have fixed the obvious cases which could lead to an inconsistent state (#103, #116)
Now what's remaining are the unknown unknowns. So I'll remove the milestone from this issue.
